kong/runloop/handler.lua

```diff
@@ -365,6 +372,190 @@ local function register_events()
+
+    for attempt = 1, max_attempts do
+      log_with_worker_id(DEBUG,
+        fmt("%s %s. Attempt %d/%d", action, name, attempt, max_attempts))
+      local pok, ok, err = pcall(callback, version)
+      if pok and ok then
+        log_with_worker_id(DEBUG, fmt("%s %s succeeded", action, name))
+        success = true
+        break
+      end
+
+      log_with_worker_id(NOTICE, fmt("%s %s failed: %s", action, name, ok or err))
+      sleep(0.01 * attempt * attempt)
+    end
+
+    if not success then
+      return nil, fmt("%s %s failed %s times, aborting",
+                      action, name, max_attempts)
+    end
+
+    return success
+  end
+
+
+  local function rebuild_timer(premature, name, callback, version, semaphore)
+    if premature then
+      release_semaphore(semaphore)
+      return
+    end
+    local ok, err = safe_rebuild(name, callback, version)
+    if not ok then
+      log(CRIT, err)
+    end
+    release_semaphore(semaphore)
+  end
+
+
+  local function rebuild_async(name, callback, version, semaphore)
+    local ok, err = timer_at(0, rebuild_timer, name, callback, version, semaphore)
+    if not ok then
+      return nil, "could not create rebuild timer: " .. err
+    end
+
+    return true
+  end
+
+
+
+  -- @param name "router" or "plugins_iterator"
+  -- @param callback A function that will update either the router or plugins_iterator
+  -- @param version target version
+  -- @param semaphore router_semaphore or plugins_iterator_semaphore
+  -- @param wait how much time to wait when trying to acquire the semaphore.
+  -- @returns true if safe_rebuild was either successfully executed synchronously,
+  -- enqueued via async timer, or not needed (because current_version == target).
+  -- nil otherwise (safe_rebuild was neither called successfully nor enqueued).
+  -- Notice that this function might return true *even when it fails to acquire
+  -- the semaphore or invoke the async callback* - it will try to rebuild
+  -- synchronously when a wait is specified and the semaphore isn't acquired
+  -- or the timer-based call fails.
+  -- @returns error message as a second return value in case of failure
+  rebuild = function(name, callback, version, semaphore, wait)
+    local current_version, err = get_version(name)
+    if not current_version then
+      return nil, err
+    end
+    if current_version == version then
+      return true
+    end
+
+    local ok = acquire_semaphore(semaphore, wait)
+    if not ok then
+      if wait and wait > 0 then
+        log(NOTICE, "could not acquire semaphore, attempting sync rebuild")
+        local ok, err = safe_rebuild(name, callback, current_version)
+        if not ok then
+          return nil, err
+        end
+        return true
+      end
+
+      return
+    end
+
+    current_version, err = get_version(name)
+    if not current_version then
+      release_semaphore(semaphore)
+      return nil, err
```

Comment:
Again, not sure if this error case should be handled as a fatal. I know it makes the interface a bit strange. My original `rebuild` didn't return anything as it handled the errors accordingly, e.g. it logged them and then just let the other code continue (it formed a border where rebuilding related errors were handled and not pushing those issues to calling code). Again, somewhat debatable, but just explaining the reasons in my original implementation, because that might not be obvious. What I mean here is that I try to follow bit different error handling than in normal code, because this is a hot path, and some errors should not in my opinion halt processing the requests. What I mean here is that we try hard (even on each following request to rebuild everything), but that is our best effort. It could be that there is a bug or maybe temporary memory issue etc., but because we cannot handle all the error cases we just try our best. Failing early just makes everything halt. I do consider there kinda of errors very severe, but not that severe, that it warrants to halt processing.

Reply:
I actually agree with part of what you are saying. This function failing should not mean always returning a bad request.

I just think that the decision of halting or not halting the request should not be done somewhere else. `rebuild` should just concern itself about rebuilding.